In [1]:
from google.colab import drive

In [2]:
drive.mount('/content/drive')

Mounted at /content/drive


#  **ETL for functions 1 and 2**

Now, with the datasets unnested and organized, we are going to do some merges and data trasnformations in order to have the information required in a CSV file to develop these functions and models:

1. **PlayTimeGenre**(`genre`: str) Function:

  It should return the release year with the most played hours for the given `genre`.

  Example of return: {"Release year with the most played hours for Genre X" : 2013}

2. **UserForGenre**(`genre`: str) Function:

  It should return the user who has accumulated the most played hours for the given `genre` and a list of the accumulated playtime by release year.

  Example of return: {"User with the most played hours for Genre X" : us213ndjss09sdf, "Playtime":[{Year: 2013, Hours: 203}, {Year: 2012, Hours: 100}, {Year: 2011, Hours: 23}]}

### 1. Data Load:

In [3]:
import pandas as pd
import numpy as np

In [4]:
df_steam_games = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/steam_games_final.csv')
df_steam_games

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

In [5]:
df_steam_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32135 entries, 0 to 32134
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   publisher     24073 non-null  object 
 1   genres        28852 non-null  object 
 2   app_name      32133 non-null  object 
 3   title         30085 non-null  object 
 4   url           32135 non-null  object 
 5   release_date  30068 non-null  object 
 6   tags          31972 non-null  object 
 7   reviews_url   32133 non-null  object 
 8   specs         31465 non-null  object 
 9   price         30758 non-null  object 
 10  early_access  32135 non-null  bool   
 11  id            32133 non-null  float64
 12  developer     28836 non-null  object 
dtypes: bool(1), float64(1), object(11)
memory usage: 3.0+ MB


In [6]:
df_users_items = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/users_items_final.csv')
df_users_items

user_id  items_count           steam_id  \
0        76561197970982479          277  76561197970982479   
1        76561197970982479          277  76561197970982479   
2        76561197970982479          277  76561197970982479   
3        76561197970982479          277  76561197970982479   
4        76561197970982479          277  76561197970982479   
...                    ...          ...                ...   
5170010  76561198329548331            7  76561198329548331   
5170011  76561198329548331            7  76561198329548331   
5170012  76561198329548331            7  76561198329548331   
5170013  76561198329548331            7  76561198329548331   
5170014  edward_tremethick            0  76561198331598578   

                                                  user_url   item_id  \
0        http://steamcommunity.com/profiles/76561197970...      10.0   
1        http://steamcommunity.com/profiles/76561197970...      20.0   
2        http://steamcommunity.com/profiles/76561197970...      30.0   
3        http://steamcommunity.com/profiles/76561197970...      40.0   
4        http://steamcommunity.com/profiles/76561197970...      50.0   
...                                                    ...       ...   
5170010  http://steamcommunity.com/profiles/76561198329...  373330.0   
5170011  http://steamcommunity.com/profiles/76561198329...  388490.0   
5170012  http://steamcommunity.com/profiles/76561198329...  521570.0   
5170013  http://steamcommunity.com/profiles/76561198329...  519140.0   
5170014     http://steamcommunity.com/id/edward_tremethick       NaN   

                             item_name  playtime_forever  playtime_2weeks  
0                       Counter-Strike               6.0              0.0  
1                Team Fortress Classic               0.0              0.0  
2                        Day of Defeat               7.0              0.0  
3                   Deathmatch Classic               0.0              0.0  
4            Half-Life: Opposing Force               0.0              0.0  
...                                ...               ...              ...  
5170010                    All Is Dust               0.0              0.0  
5170011  One Way To Die: Steam Edition               3.0              3.0  
5170012          You Have 10 Seconds 2               4.0              4.0  
5170013                     Minds Eyes               3.0              3.0  
5170014                            NaN               NaN              NaN  

[5170015 rows x 8 columns]

In [7]:
df_users_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5170015 entries, 0 to 5170014
Data columns (total 8 columns):
 #   Column            Dtype  
---  ------            -----  
 0   user_id           object 
 1   items_count       int64  
 2   steam_id          int64  
 3   user_url          object 
 4   item_id           float64
 5   item_name         object 
 6   playtime_forever  float64
 7   playtime_2weeks   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 315.6+ MB


**Step 1:**

df_steam_games has a column called `release_date`. We need to extract the year of this release date and place it in a new column called `release_year`. We ar going to do something similar as in the step 1.

In [8]:
df_steam_games

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

As in the step 1, it is a good practice to check the NaN in the column 'release_date'.

In [9]:
mask = df_steam_games['release_date'].isna()
df_steam_games[mask]

publisher                                            genres  \
4                   NaN                                               NaN   
11                  NaN                                               NaN   
19                  NaN                                               NaN   
20                  NaN                                               NaN   
22                  NaN                                               NaN   
...                 ...                                               ...   
32077               NaN                                               NaN   
32084               NaN                     ['Action', 'Casual', 'Indie']   
32085               NaN                                               NaN   
32088  rechargecomplete  ['Action', 'Adventure', 'Indie', 'Early Access']   
32134               NaN                                               NaN   

                   app_name              title  \
4             Log Challenge                NaN   
11     Icarus Six Sixty Six                NaN   
19            After Life VR                NaN   
20               Kitty Hawk                NaN   
22               Mortars VR                NaN   
...                     ...                ...   
32077           SpaceWalker                NaN   
32084              BOMBFEST           BOMBFEST   
32085            LIV Client                NaN   
32088     RECHARGE COMPLETE  RECHARGE COMPLETE   
32134           Maze Run VR                NaN   

                                                     url release_date  \
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
11     http://store.steampowered.com/app/724910/Icaru...          NaN   
19     http://store.steampowered.com/app/772590/After...          NaN   
20     http://store.steampowered.com/app/640250/Kitty...          NaN   
22     http://store.steampowered.com/app/711440/Morta...          NaN   
...                                                  ...          ...   
32077  http://store.steampowered.com/app/705860/Space...          NaN   
32084  http://store.steampowered.com/app/729640/BOMBF...          NaN   
32085  http://store.steampowered.com/app/755540/LIV_C...          NaN   
32088  http://store.steampowered.com/app/708070/RECHA...          NaN   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
4                ['Action', 'Indie', 'Casual', 'Sports']   
11                                            ['Casual']   
19                       ['Early Access', 'Indie', 'VR']   
20     ['Early Access', 'Action', 'Adventure', 'Indie...   
22     ['Early Access', 'Strategy', 'Action', 'Indie'...   
...                                                  ...   
32077                         ['Early Access', 'Casual']   
32084                      ['Casual', 'Indie', 'Action']   
32085  ['Video Production', 'Utilities', 'Web Publish...   
32088   ['Early Access', 'Action', 'Adventure', 'Indie']   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
4      http://steamcommunity.com/app/773570/reviews/?...   
11     http://steamcommunity.com/app/724910/reviews/?...   
19     http://steamcommunity.com/app/772590/reviews/?...   
20     http://steamcommunity.com/app/640250/reviews/?...   
22     http://steamcommunity.com/app/711440/reviews/?...   
...                                                  ...   
32077  http://steamcommunity.com/app/705860/reviews/?...   
32084  http://steamcommunity.com/app/729640/reviews/?...   
32085  http://steamcommunity.com/app/755540/reviews/?...   
32088  http://steamcommunity.com/app/708070/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs price  early_access  \
4      ['Single-player', 'Full controller support', '...  2.99         False   
11     ['

In this case, we can see that though the 'release_date' is in NaN, the entries have other information that can be useful.

Now, let's review the format of the dates in the column 'release_date', to check how can we extract the year.

In [10]:
df_steam_games['release_date'][:60]

0     2018-01-04
1     2018-01-04
2     2017-07-24
3     2017-12-07
4            NaN
5     2018-01-04
6     2018-01-04
7     2018-01-04
8     2018-01-04
9     2018-01-04
10        Soon..
11           NaN
12    2018-01-04
13    2018-01-04
14    2018-01-03
15    2018-01-03
16    2018-01-03
17    2018-01-03
18    2018-01-03
19           NaN
20           NaN
21    2018-01-03
22           NaN
23    2017-12-22
24    2017-12-23
25    2018-01-03
26    2018-01-03
27    1997-06-30
28    1998-11-08
29    2016-11-25
30           NaN
31    2018-01-01
32    2018-01-01
33    2018-01-01
34    2018-01-01
35    2017-12-30
36           NaN
37    2018-01-01
38    2018-01-01
39    2006-07-06
40    2006-07-06
41    2006-07-11
42           NaN
43          2017
44     Beta测试已开启
45           NaN
46    2017-12-29
47    2018-03-30
48    2017-12-29
49    2017-12-29
50    2017-12-29
51           NaN
52    2017-12-30
53           NaN
54    2017-12-29
55    2005-08-09
56    2006-09-29
57    2006-11-20
58    2006-11-

In the sample of the first 60 rows we can see that the normal format is YYYY-MM-DD. However, there are some NaN and strings that need to be considered. Also, there are entries with only the year. So we need to consider also this.

In [11]:
import re

# We create a function to extract the year considering the data that we have in the sample above.
def extract_year(date_str):
    if pd.notna(date_str):  #If the value is not NaN
        year_match = re.search(r'\b\d{4}\b', date_str) # We loo for the year following the format '\b\d{4}\b' using the function re.search()
        if year_match:
            return year_match.group()
    return np.NaN

# we apply the function to the column release_date
release_year = df_steam_games['release_date'].apply(extract_year)

# Mostrar el DataFrame resultante
release_year

0        2018
1        2018
2        2017
3        2017
4         NaN
         ... 
32130    2018
32131    2018
32132    2018
32133    2017
32134     NaN
Name: release_date, Length: 32135, dtype: object

Now we concatenate the column release_year with the dataframe steam_games:

In [12]:
df_steam_games_WY = pd.concat([df_steam_games, pd.DataFrame({'release_year': release_year})], axis=1)
df_steam_games_WY

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

**Step 3:**

Having a dataframe with the item IDs for that specific `genre`, we can merge it with a dataframe that has the sum of 'playtime_forever' grouped by `item_id`. This information is found in the dataframe df_users_items. Now we have a new dataframe for a specific genre with the sum of hours played.

In [13]:
df_users_items

user_id  items_count           steam_id  \
0        76561197970982479          277  76561197970982479   
1        76561197970982479          277  76561197970982479   
2        76561197970982479          277  76561197970982479   
3        76561197970982479          277  76561197970982479   
4        76561197970982479          277  76561197970982479   
...                    ...          ...                ...   
5170010  76561198329548331            7  76561198329548331   
5170011  76561198329548331            7  76561198329548331   
5170012  76561198329548331            7  76561198329548331   
5170013  76561198329548331            7  76561198329548331   
5170014  edward_tremethick            0  76561198331598578   

                                                  user_url   item_id  \
0        http://steamcommunity.com/profiles/76561197970...      10.0   
1        http://steamcommunity.com/profiles/76561197970...      20.0   
2        http://steamcommunity.com/profiles/76561197970...      30.0   
3        http://steamcommunity.com/profiles/76561197970...      40.0   
4        http://steamcommunity.com/profiles/76561197970...      50.0   
...                                                    ...       ...   
5170010  http://steamcommunity.com/profiles/76561198329...  373330.0   
5170011  http://steamcommunity.com/profiles/76561198329...  388490.0   
5170012  http://steamcommunity.com/profiles/76561198329...  521570.0   
5170013  http://steamcommunity.com/profiles/76561198329...  519140.0   
5170014     http://steamcommunity.com/id/edward_tremethick       NaN   

                             item_name  playtime_forever  playtime_2weeks  
0                       Counter-Strike               6.0              0.0  
1                Team Fortress Classic               0.0              0.0  
2                        Day of Defeat               7.0              0.0  
3                   Deathmatch Classic               0.0              0.0  
4            Half-Life: Opposing Force               0.0              0.0  
...                                ...               ...              ...  
5170010                    All Is Dust               0.0              0.0  
5170011  One Way To Die: Steam Edition               3.0              3.0  
5170012          You Have 10 Seconds 2               4.0              4.0  
5170013                     Minds Eyes               3.0              3.0  
5170014                            NaN               NaN              NaN  

[5170015 rows x 8 columns]

Let's group by item_id summing the playtime_forever column. To so this, we can use the function "groupby()" of Pandas:

In [14]:
# Group by "item_id" and sum the values of "playtime_forever"
grouped_df_users_items = df_users_items.groupby(['item_id', 'user_id'])['playtime_forever'].sum().reset_index()

grouped_df_users_items.rename(columns={'playtime_forever': 'sum_playtime_forever'}, inplace=True)
grouped_df_users_items

item_id               user_id  sum_playtime_forever
0            10.0            -GM-Dragon                  14.0
1            10.0            -KillZone-                 987.0
2            10.0                 -Mad-                   8.0
3            10.0               -SEVEN-                 617.0
4            10.0  -SatansLittleHelper-                 511.0
...           ...                   ...                   ...
5094077  529670.0                chidvd                   0.0
5094078  529670.0        fearlesskeeper                   0.0
5094079  529820.0           robinwalker                   0.0
5094080  530720.0            CSMisBeast                   0.0
5094081  530720.0        mayshowganmore                   0.0

[5094082 rows x 3 columns]

We passed from 5,170,015 entries to 5,094,082 registers.

Now we need to merge the grouped dataframe grouped_df_users_items with the df_steam_games_WY in order to have all the information together and apply the functions:

In [15]:
df_steam_games_WY.rename(columns={'id': 'item_id'}, inplace=True)
df_steam_games_WY

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

In [16]:
#Merging the dataframes by item_id.
merged_all_by_sumPTForever = pd.merge(grouped_df_users_items, df_steam_games_WY, on='item_id', how='inner')
merged_all_by_sumPTForever

item_id               user_id  sum_playtime_forever  \
0            10.0            -GM-Dragon                  14.0   
1            10.0            -KillZone-                 987.0   
2            10.0                 -Mad-                   8.0   
3            10.0               -SEVEN-                 617.0   
4            10.0  -SatansLittleHelper-                 511.0   
...           ...                   ...                   ...   
4244807  528660.0            SirexPower                 182.0   
4244808  528660.0          thugnificent                   0.0   
4244809  529820.0           robinwalker                   0.0   
4244810  530720.0            CSMisBeast                   0.0   
4244811  530720.0        mayshowganmore                   0.0   

                  publisher                         genres          app_name  \
0                     Valve                     ['Action']    Counter-Strike   
1                     Valve                     ['Action']    Counter-Strike   
2                     Valve                     ['Action']    Counter-Strike   
3                     Valve                     ['Action']    Counter-Strike   
4                     Valve                     ['Action']    Counter-Strike   
...                     ...                            ...               ...   
4244807  Just1337 Publisher  ['Action', 'Casual', 'Indie']         Spin Rush   
4244808  Just1337 Publisher  ['Action', 'Casual', 'Indie']         Spin Rush   
4244809                 NaN                            NaN   Beyond Power VR   
4244810          EnsenaSoft            ['Casual', 'Indie']  Solitaire Royale   
4244811          EnsenaSoft            ['Casual', 'Indie']  Solitaire Royale   

                    title                                                url  \
0          Counter-Strike  http://store.steampowered.com/app/10/CounterSt...   
1          Counter-Strike  http://store.steampowered.com/app/10/CounterSt...   
2          Counter-Strike  http://store.steampowered.com/app/10/CounterSt...   
3          Counter-Strike  http://store.steampowered.com/app/10/CounterSt...   
4          Counter-Strike  http://store.steampowered.com/app/10/CounterSt...   
...                   ...                                                ...   
4244807         Spin Rush  http://store.steampowered.com/app/528660/Spin_...   
4244808         Spin Rush  http://store.steampowered.com/app/528660/Spin_...   
4244809               NaN  http://store.steampowered.com/app/529820/Beyon...   
4244810  Solitaire Royale  http://store.steampowered.com/app/530720/Solit...   
4244811  Solitaire Royale  http://store.steampowered.com/app/530720/Solit...   

        release_date                                               tags  \
0         2000-11-01  ['Action', 'FPS', 'Multiplayer', 'Shooter', 'C...   
1         2000-11-01  ['Action', 'FPS', 'Multiplayer', 'Shooter', 'C...   
2         2000-11-01  ['Action', 'FPS', 'Multiplayer', 'Shooter', 'C...   
3         2000-11-01  ['Action', 'FPS', 'Multiplayer', 'Shooter', 'C...   
4         2000-11-01  ['Action', 'FPS', 'Multiplayer', 'Shooter', 'C...   
...              ...                                                ...   
4244807   2016-09-15            ['Indie', 'Casual', 'Action', 'Puzzle']   
4244808   2016-09-15            ['Indie', 'Casual', 'Action', 'Puzzle']   
4244809          NaN                          ['Action', 'Indie', 'VR']   
4244810   2016-09-15                                ['Casual', 'Indie']   
4244811   2016-09-15                                ['Casual', 'Indie']   

                                               reviews_url  \
0        http://steamcommunity.com/app/10/reviews/?brow...   
1        http://steamcommunity.com/app/10/reviews/?brow...   
2        http://steamcommunity.com/app/10/reviews/?brow...   
3        http://steamcommunity.com/app/10/reviews/?brow...   
4        http://steamcommunity.com/app/10/reviews/?brow...   
...                               

In [17]:
merged_all_by_sumPTForever.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4244812 entries, 0 to 4244811
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   item_id               float64
 1   user_id               object 
 2   sum_playtime_forever  float64
 3   publisher             object 
 4   genres                object 
 5   app_name              object 
 6   title                 object 
 7   url                   object 
 8   release_date          object 
 9   tags                  object 
 10  reviews_url           object 
 11  specs                 object 
 12  price                 object 
 13  early_access          bool   
 14  developer             object 
 15  release_year          object 
dtypes: bool(1), float64(2), object(13)
memory usage: 522.2+ MB


Now we need to keep minimum columns that are requiered by the functions in order to save resources.

Analyzing the functions, the only columns that we are going to need are the following:

- user_id: We need these data for the function 2.
- item_id: It is important for the ML model where the user needs to enter an item_id in order to receive 5 games that are similar to the one selected.
- title: It is important for the ML model where the user needs to enter an item_id in order to receive 5 games that are similar to the one selected.
- release_year: We need these data for the function 1 and 2.
- genres: We need these data for the function 1 and 2.
- sum_playtime_forever: We need these data for the function 1 and 2.

In [46]:
selected_columns = ['user_id','item_id','title','release_year','genres','sum_playtime_forever']
df_F12 = merged_all_by_sumPTForever[selected_columns]
df_F12

user_id   item_id             title release_year  \
0                  -GM-Dragon      10.0    Counter-Strike         2000   
1                  -KillZone-      10.0    Counter-Strike         2000   
2                       -Mad-      10.0    Counter-Strike         2000   
3                     -SEVEN-      10.0    Counter-Strike         2000   
4        -SatansLittleHelper-      10.0    Counter-Strike         2000   
...                       ...       ...               ...          ...   
4244807            SirexPower  528660.0         Spin Rush         2016   
4244808          thugnificent  528660.0         Spin Rush         2016   
4244809           robinwalker  529820.0               NaN          NaN   
4244810            CSMisBeast  530720.0  Solitaire Royale         2016   
4244811        mayshowganmore  530720.0  Solitaire Royale         2016   

                                genres  sum_playtime_forever  
0                           ['Action']                  14.0  
1                           ['Action']                 987.0  
2                           ['Action']                   8.0  
3                           ['Action']                 617.0  
4                           ['Action']                 511.0  
...                                ...                   ...  
4244807  ['Action', 'Casual', 'Indie']                 182.0  
4244808  ['Action', 'Casual', 'Indie']                   0.0  
4244809                            NaN                   0.0  
4244810            ['Casual', 'Indie']                   0.0  
4244811            ['Casual', 'Indie']                   0.0  

[4244812 rows x 6 columns]

In order to reduce the size of the dataframe we can eliminate the rows where genres are NaN, the sum_playtime_forever is 0.0 or the release_year is NaN.

First, let's check if there is any user in NaN:

In [47]:
mask = df_F12['user_id'].isna()
df_F12[mask]

Empty DataFrame
Columns: [user_id, item_id, title, release_year, genres, sum_playtime_forever]
Index: []

We can see that there is no user with NaN, so any NaN in the other remaining columns would eliminate the row to be considered by the functions. So, in this case, we can apply the function dropna() to all the dataframe:

In [48]:
df_F12.dropna(inplace=True)
df_F12

<ipython-input-48-16c7fcd4b8bb>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_F12.dropna(inplace=True)


user_id   item_id  \
0                  -GM-Dragon      10.0   
1                  -KillZone-      10.0   
2                       -Mad-      10.0   
3                     -SEVEN-      10.0   
4        -SatansLittleHelper-      10.0   
...                       ...       ...   
4244800     76561198159037032  527900.0   
4244807            SirexPower  528660.0   
4244808          thugnificent  528660.0   
4244810            CSMisBeast  530720.0   
4244811        mayshowganmore  530720.0   

                                             title release_year  \
0                                   Counter-Strike         2000   
1                                   Counter-Strike         2000   
2                                   Counter-Strike         2000   
3                                   Counter-Strike         2000   
4                                   Counter-Strike         2000   
...                                            ...          ...   
4244800  Grim Tales: The Bride Collector's Edition         2016   
4244807                                  Spin Rush         2016   
4244808                                  Spin Rush         2016   
4244810                           Solitaire Royale         2016   
4244811                           Solitaire Royale         2016   

                                genres  sum_playtime_forever  
0                           ['Action']                  14.0  
1                           ['Action']                 987.0  
2                           ['Action']                   8.0  
3                           ['Action']                 617.0  
4                           ['Action']                 511.0  
...                                ...                   ...  
4244800        ['Adventure', 'Casual']                  44.0  
4244807  ['Action', 'Casual', 'Indie']                 182.0  
4244808  ['Action', 'Casual', 'Indie']                   0.0  
4244810            ['Casual', 'Indie']                   0.0  
4244811            ['Casual', 'Indie']                   0.0  

[4149842 rows x 6 columns]

Finally, we need to eliminate the sum_playtime_forever in 0.0. These rows do not have any value for the functions:

In [49]:
mask = df_F12['sum_playtime_forever']!= 0.0
df_F12 = df_F12[mask]
df_F12

user_id   item_id  \
0                  -GM-Dragon      10.0   
1                  -KillZone-      10.0   
2                       -Mad-      10.0   
3                     -SEVEN-      10.0   
4        -SatansLittleHelper-      10.0   
...                       ...       ...   
4244776        mayshowganmore  527510.0   
4244793                 lexby  527810.0   
4244799          SmokaBluntos  527890.0   
4244800     76561198159037032  527900.0   
4244807            SirexPower  528660.0   

                                             title release_year  \
0                                   Counter-Strike         2000   
1                                   Counter-Strike         2000   
2                                   Counter-Strike         2000   
3                                   Counter-Strike         2000   
4                                   Counter-Strike         2000   
...                                            ...          ...   
4244776                        The Legions of Rome         2016   
4244793                              Dynamite Alex         2016   
4244799            Shop-n-Spree: Shopping Paradise         2016   
4244800  Grim Tales: The Bride Collector's Edition         2016   
4244807                                  Spin Rush         2016   

                                                genres  sum_playtime_forever  
0                                           ['Action']                  14.0  
1                                           ['Action']                 987.0  
2                                           ['Action']                   8.0  
3                                           ['Action']                 617.0  
4                                           ['Action']                 511.0  
...                                                ...                   ...  
4244776  ['Casual', 'Indie', 'Simulation', 'Strategy']                 154.0  
4244793                            ['Action', 'Indie']                   2.0  
4244799                       ['Casual', 'Simulation']                   1.0  
4244800                        ['Adventure', 'Casual']                  44.0  
4244807                  ['Action', 'Casual', 'Indie']                 182.0  

[2763977 rows x 6 columns]

In [50]:
df_F12.reset_index(drop=True, inplace=True)
df_F12

user_id   item_id  \
0                  -GM-Dragon      10.0   
1                  -KillZone-      10.0   
2                       -Mad-      10.0   
3                     -SEVEN-      10.0   
4        -SatansLittleHelper-      10.0   
...                       ...       ...   
2763972        mayshowganmore  527510.0   
2763973                 lexby  527810.0   
2763974          SmokaBluntos  527890.0   
2763975     76561198159037032  527900.0   
2763976            SirexPower  528660.0   

                                             title release_year  \
0                                   Counter-Strike         2000   
1                                   Counter-Strike         2000   
2                                   Counter-Strike         2000   
3                                   Counter-Strike         2000   
4                                   Counter-Strike         2000   
...                                            ...          ...   
2763972                        The Legions of Rome         2016   
2763973                              Dynamite Alex         2016   
2763974            Shop-n-Spree: Shopping Paradise         2016   
2763975  Grim Tales: The Bride Collector's Edition         2016   
2763976                                  Spin Rush         2016   

                                                genres  sum_playtime_forever  
0                                           ['Action']                  14.0  
1                                           ['Action']                 987.0  
2                                           ['Action']                   8.0  
3                                           ['Action']                 617.0  
4                                           ['Action']                 511.0  
...                                                ...                   ...  
2763972  ['Casual', 'Indie', 'Simulation', 'Strategy']                 154.0  
2763973                            ['Action', 'Indie']                   2.0  
2763974                       ['Casual', 'Simulation']                   1.0  
2763975                        ['Adventure', 'Casual']                  44.0  
2763976                  ['Action', 'Casual', 'Indie']                 182.0  

[2763977 rows x 6 columns]

Now we have in total 2,763,977 entries and 4 rows, which is the only data that we need to solve the functions.

In [51]:
df_F12.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2763977 entries, 0 to 2763976
Data columns (total 6 columns):
 #   Column                Dtype  
---  ------                -----  
 0   user_id               object 
 1   item_id               float64
 2   title                 object 
 3   release_year          object 
 4   genres                object 
 5   sum_playtime_forever  float64
dtypes: float64(2), object(4)
memory usage: 126.5+ MB


We need to change the "title" type from object to string

In [52]:
df_F12['title'] = df_F12['title'].astype('string')
df_F12.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2763977 entries, 0 to 2763976
Data columns (total 6 columns):
 #   Column                Dtype  
---  ------                -----  
 0   user_id               object 
 1   item_id               float64
 2   title                 string 
 3   release_year          object 
 4   genres                object 
 5   sum_playtime_forever  float64
dtypes: float64(2), object(3), string(1)
memory usage: 126.5+ MB


<ipython-input-52-9bfb76e8eadd>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_F12['title'] = df_F12['title'].astype('string')


We need to change the "release_year" type from object to int64

In [53]:
df_F12['release_year'] = df_F12['release_year'].astype('int64')
df_F12.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2763977 entries, 0 to 2763976
Data columns (total 6 columns):
 #   Column                Dtype  
---  ------                -----  
 0   user_id               object 
 1   item_id               float64
 2   title                 string 
 3   release_year          int64  
 4   genres                object 
 5   sum_playtime_forever  float64
dtypes: float64(2), int64(1), object(2), string(1)
memory usage: 126.5+ MB


<ipython-input-53-731ca712363f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_F12['release_year'] = df_F12['release_year'].astype('int64')


In [54]:
duplicados_item_id = df_F12['item_id'].duplicated()
df_F12[duplicados_item_id]

user_id   item_id                title  release_year  \
1                  -KillZone-      10.0       Counter-Strike          2000   
2                       -Mad-      10.0       Counter-Strike          2000   
3                     -SEVEN-      10.0       Counter-Strike          2000   
4        -SatansLittleHelper-      10.0       Counter-Strike          2000   
5                     -Ultrix      10.0       Counter-Strike          2000   
...                       ...       ...                  ...           ...   
2763967            kingholica  527340.0        What The Box?          2016   
2763968               livaris  527340.0        What The Box?          2016   
2763969             sky_track  527340.0        What The Box?          2016   
2763970       theghostofbowie  527340.0        What The Box?          2016   
2763972        mayshowganmore  527510.0  The Legions of Rome          2016   

                                                genres  sum_playtime_forever  
1                                           ['Action']                 987.0  
2                                           ['Action']                   8.0  
3                                           ['Action']                 617.0  
4                                           ['Action']                 511.0  
5                                           ['Action']                   5.0  
...                                                ...                   ...  
2763967                            ['Action', 'Indie']                  89.0  
2763968                            ['Action', 'Indie']                 176.0  
2763969                            ['Action', 'Indie']                   6.0  
2763970                            ['Action', 'Indie']                  12.0  
2763972  ['Casual', 'Indie', 'Simulation', 'Strategy']                 154.0  

[2755882 rows x 6 columns]

For the functions 1 and 2, we only require these 4 columns:

- user_id
- release_year
- genres
- sum_playtime_forever

In [56]:
columns_to_eliminate = ['item_id','title']
df_F12 = df_F12.drop(columns=columns_to_eliminate)
df_F12


user_id  release_year  \
0                  -GM-Dragon          2000   
1                  -KillZone-          2000   
2                       -Mad-          2000   
3                     -SEVEN-          2000   
4        -SatansLittleHelper-          2000   
...                       ...           ...   
2763972        mayshowganmore          2016   
2763973                 lexby          2016   
2763974          SmokaBluntos          2016   
2763975     76561198159037032          2016   
2763976            SirexPower          2016   

                                                genres  sum_playtime_forever  
0                                           ['Action']                  14.0  
1                                           ['Action']                 987.0  
2                                           ['Action']                   8.0  
3                                           ['Action']                 617.0  
4                                           ['Action']                 511.0  
...                                                ...                   ...  
2763972  ['Casual', 'Indie', 'Simulation', 'Strategy']                 154.0  
2763973                            ['Action', 'Indie']                   2.0  
2763974                       ['Casual', 'Simulation']                   1.0  
2763975                        ['Adventure', 'Casual']                  44.0  
2763976                  ['Action', 'Casual', 'Indie']                 182.0  

[2763977 rows x 4 columns]

Now we need to group by user_id summing sum_playtime_forever.

In [83]:
groupedF12 = df_F12.groupby(['user_id', 'release_year', 'genres'])['sum_playtime_forever'].sum().reset_index()
groupedF12

user_id  release_year  \
0           --000--          2006   
1           --000--          2009   
2           --000--          2010   
3           --000--          2011   
4           --000--          2011   
...             ...           ...   
2241396  zzzmidmiss          2012   
2241397  zzzmidmiss          2013   
2241398  zzzmidmiss          2013   
2241399  zzzmidmiss          2013   
2241400  zzzmidmiss          2014   

                                                    genres  \
0                                  ['Indie', 'Simulation']   
1                                               ['Action']   
2                      ['Action', 'Free to Play', 'Indie']   
3                  ['Action', 'Adventure', 'Indie', 'RPG']   
4        ['Action', 'Free to Play', 'Massively Multipla...   
...                                                    ...   
2241396                                            ['RPG']   
2241397            ['Action', 'Adventure', 'Indie', 'RPG']   
2241398                         ['Action', 'Free to Play']   
2241399                   ['Adventure', 'Casual', 'Indie']   
2241400  ['Action', 'Adventure', 'Free to Play', 'Simul...   

         sum_playtime_forever  
0                       925.0  
1                      5329.0  
2                        22.0  
3                      1172.0  
4                      2763.0  
...                       ...  
2241396                  93.0  
2241397                  88.0  
2241398                  10.0  
2241399                  17.0  
2241400                  16.0  

[2241401 rows x 4 columns]

We can see that there are some repeated release_years with the same genres. We need to group by the release_year and the user_id concatenating the genres and summing the sum_playtime_forever. Let's do it:

In [125]:
# We define a function to convert the string representation to a list format
def convert_to_list_format(s):
    try:
        # Remove brackets, split by ', ', and enclose each element in single quotes
        elements = s.strip('[]').split(', ')
        formatted_elements = [f"'{element}'" for element in elements]
        return formatted_elements
    except Exception as e:
        # Handle any exceptions that may occur during conversion
        return []

# We apply the conversion function to the 'genres' column
groupedF12['genres'] = groupedF12['genres'].apply(convert_to_list_format)

# Then, explode the 'genres' column so that each genre is in a separate row
groupedF12 = groupedF12.explode('genres')

# Then, group by 'user_id' and 'release_year', concatenating the genres and summing the hours
aggregated_df = groupedF12.groupby(['user_id', 'release_year']).agg({'genres': list, 'sum_playtime_forever': 'sum'}).reset_index()

# Now we have a DataFrame with genres in the format ['Genre1', 'Genre2']
aggregated_df

user_id  release_year  \
0          --000--          2006   
1          --000--          2009   
2          --000--          2010   
3          --000--          2011   
4          --000--          2012   
...            ...           ...   
660198  zzzmidmiss          2010   
660199  zzzmidmiss          2011   
660200  zzzmidmiss          2012   
660201  zzzmidmiss          2013   
660202  zzzmidmiss          2014   

                                                   genres  \
0                             [''Indie'', ''Simulation'']   
1                                            [''Action'']   
2               [''Action'', ''Free to Play'', ''Indie'']   
3       [''Action'', ''Adventure'', ''Indie'', ''RPG''...   
4       [''Action'', ''Adventure'', ''Casual'', ''Indi...   
...                                                   ...   
660198  [''Action'', ''Free to Play'', ''Indie'', ''Ad...   
660199  [''Action'', ''Free to Play'', ''Massively Mul...   
660200  [''Action'', ''Adventure'', ''Action'', ''Free...   
660201  [''Action'', ''Adventure'', ''Indie'', ''RPG''...   
660202  [''Action'', ''Adventure'', ''Free to Play'', ...   

        sum_playtime_forever  
0                     1850.0  
1                     5329.0  
2                       66.0  
3                    15564.0  
4                   118326.0  
...                      ...  
660198                1237.0  
660199                2514.0  
660200               14467.0  
660201                 423.0  
660202                  80.0  

[660203 rows x 4 columns]

In [126]:
import pandas as pd
import numpy as np

# Suppose aggregated_df is your DataFrame
# Define a function to convert the values
def convert_to_list(genres_array):
    return [element.strip("'") for element in genres_array]

# Apply the function to the 'genres' column and replace the column
aggregated_df['genres'] = aggregated_df['genres'].apply(convert_to_list)

aggregated_df

user_id  release_year  \
0          --000--          2006   
1          --000--          2009   
2          --000--          2010   
3          --000--          2011   
4          --000--          2012   
...            ...           ...   
660198  zzzmidmiss          2010   
660199  zzzmidmiss          2011   
660200  zzzmidmiss          2012   
660201  zzzmidmiss          2013   
660202  zzzmidmiss          2014   

                                                   genres  \
0                                     [Indie, Simulation]   
1                                                [Action]   
2                           [Action, Free to Play, Indie]   
3       [Action, Adventure, Indie, RPG, Action, Free t...   
4       [Action, Adventure, Casual, Indie, RPG, Action...   
...                                                   ...   
660198  [Action, Free to Play, Indie, Adventure, Strat...   
660199  [Action, Free to Play, Massively Multiplayer, ...   
660200  [Action, Adventure, Action, Free to Play, Indi...   
660201  [Action, Adventure, Indie, RPG, Action, Free t...   
660202  [Action, Adventure, Free to Play, Simulation, ...   

        sum_playtime_forever  
0                     1850.0  
1                     5329.0  
2                       66.0  
3                    15564.0  
4                   118326.0  
...                      ...  
660198                1237.0  
660199                2514.0  
660200               14467.0  
660201                 423.0  
660202                  80.0  

[660203 rows x 4 columns]

We are going to save the entire dataset in a CSV file to have a backup copy ready to be consumed in subsequent processes.

In [127]:
aggregated_df.to_csv('F12.csv', index=False)

Let's save it also in a parquet file due to it saves resources.

In [128]:
aggregated_df.to_parquet('F12.parquet', index=False)